src/transaction_manager/transaction_queue.cpp

```diff
@@ -20,45 +20,80 @@ void TransactionQueue::stop() {
   cond_for_unverified_qu_.notify_all();
 }
 
-void TransactionQueue::insert(Transaction const &trx, bool verify) {
-  trx_hash_t hash = trx.getHash();
+bool TransactionQueue::insert(Transaction const &trx, bool queue_verified, bool tx_verified,
+                              const std::shared_ptr<DbStorage> &db) {
+  const auto hash = trx.getHash();
   auto trx_ptr = make_shared<Transaction>(trx);
+
   {
     uLock lock(main_shared_mutex_);
-    queued_trxs_[hash] = trx_ptr;
-    if (verify) {
-      verified_trxs_[hash] = trx_ptr;
+
+    // Transaction is already in queued_trxs_, do not insert it again.
+    if (!queued_trxs_.emplace(hash, trx_ptr).second) {
+      LOG(log_nf_) << "Tx " << hash.abridged() << " already inserted, skip it";
+      return false;
+    }
+
+    TransactionStatusEnum tx_status = TransactionStatusEnum::in_queue_unverified;
+
+    // shared_mutex_for_unverified_qu_ moved out of "if (queue_verified)" else branch so the lock is released
+    // after db operations are performed. Otherwise there is race condition
+    uLock unlock(shared_mutex_for_unverified_qu_);
```

Comment:
I do not understand how there is race condition as `shared_mutex_for_unverified_qu_` is used only when touching `unverified_hash_qu_` and not DB

Reply:
this is "logic" kind of race condition where we need to have sort of atomic change for unverified_hash_qu_ + db